test/test_lib.py

```diff
@@ -422,91 +409,108 @@ def test_use_finds_subs(self):
         assert baz.LIBPATCH == 42
         assert baz.LIBAUTHOR == 'alice@example.com'
 
-    def test_use_finds_best_same_toplevel(self):
+    @pytest.mark.parametrize("pkg_a", ["foo", "fooA"])
+    @pytest.mark.parametrize("lib_a", ["bar", "barA"])
+    @pytest.mark.parametrize("patch_a", [38, 42])
+    def test_use_finds_best_same_toplevel(
+        self,
+        tmp_path: pathlib.Path,
+        monkeypatch: pytest.MonkeyPatch,
+        pkg_a: str,
+        lib_a: str,
+        patch_a: int
+    ):
         """Test that ops.lib.use("baz") works when there are two baz in the same toplevel."""
         pkg_b = "foo"
         lib_b = "bar"
         patch_b = 40
-        for pkg_a in ["foo", "fooA"]:
-            for lib_a in ["bar", "barA"]:
-                if (pkg_a, lib_a) == (pkg_b, lib_b):
-                    # everything-is-the-same :-)
-                    continue
-                for patch_a in [38, 42]:
-                    desc = f"A: {pkg_a}/{lib_a}/{patch_a}; B: {pkg_b}/{lib_b}/{patch_b}"
-                    with self.subTest(desc):
-                        tmpdir = self._mkdtemp()
-                        sys.path = [tmpdir]
-
-                        _mklib(tmpdir, pkg_a, lib_a).write_text(dedent("""
-                        LIBNAME = "baz"
-                        LIBAPI = 2
-                        LIBPATCH = {}
-                        LIBAUTHOR = "alice@example.com"
-                        """).format(patch_a))
-
-                        _mklib(tmpdir, pkg_b, lib_b).write_text(dedent("""
-                        LIBNAME = "baz"
-                        LIBAPI = 2
-                        LIBPATCH = {}
-                        LIBAUTHOR = "alice@example.com"
-                        """).format(patch_b))
-
-                        # autoimport to reset things
-                        ops.lib.autoimport()
-
-                        # ops.lib.use done by charm author
-                        baz = ops.lib.use('baz', 2, 'alice@example.com')
-                        assert baz.LIBNAME == 'baz'
-                        assert baz.LIBAPI == 2
-                        assert max(patch_a, patch_b) == baz.LIBPATCH
-                        assert baz.LIBAUTHOR == 'alice@example.com'
-
-    def test_use_finds_best_diff_toplevel(self):
+
+        if (pkg_a, lib_a) == (pkg_b, lib_b):
+            pytest.skip('Invalid case')
```

Comment:
It doesn't seem great to have tests that will always be skipped. Is there some way we can adjust the parametrizing to avoid this?

Reply:
I will return early then.

Reply:
A test that always passes is better than one that always skips, I think. It does seem like just adjusting the parametrizing so that only the valid cases are run would probably be better, but I think this is fine for now.